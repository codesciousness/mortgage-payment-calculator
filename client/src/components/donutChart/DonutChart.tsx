import { relative } from 'path';
import React from 'react';
import { Chart } from "react-google-charts";
import { styles } from '../../styles';

enum Expenses {
    PI = 'Principal & interest',
    PT = 'Property tax',
    HI = 'Homeowner\'s insurance',
    HOA = 'HOA fees',
    OTH = 'Other costs'
};

type ColumnHeaders = [string, string];

type RowData = [Expenses, number];

const DonutChart = (): JSX.Element => {
    const columnHeaders: ColumnHeaders = ['Mortgage Related Expenses', 'Payment Amount'];
    const rowData: RowData = [
        [Expenses.PI, 11],
        [Expenses.PT, 2],
        [Expenses.HI, 2],
        [Expenses.HOA, 2],
        [Expenses.OTH, 7]
    ];
    const data = [columnHeaders, ...rowData];

    const total = '$2,500';

    const options = {
        pieHole: 0.6,
        is3D: false,
        pieSliceText: 'none',
        legend: 'none',
        backgroundColor: 'whitesmoke',
        slices: {
            0: { color: 'DeepSkyBlue' },
            1: { color: 'LightSkyBlue' },
            2: { color: 'SkyBlue' },
            3: { color: 'PowderBlue' },
            4: { color: 'PaleTurquoise' }
        }
    };

    return (
        <table id='DonutChart'>
            <tr>
                <td style={styles.chartTableData}>
                    <Chart
                        chartType='PieChart'
                        width='500px'
                        height='500px'
                        data={data}
                        options={options}
                    />
                    <div id='Overlay' style={styles.chartOverlay}>
                        <p style={styles.chartOverlayOtherText}>Monthly Payment</p>
                        <p style={styles.chartOverlayMainText}>{total}</p>
                    </div>
                </td>
            </tr>
        </table>
    );
};

export default DonutChart;