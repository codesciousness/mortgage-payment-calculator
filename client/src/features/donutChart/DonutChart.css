.DonutChart__tableData {
    display: flex;
    justify-content: center;
    align-items: center;
    position: relative;
}

.DonutChart__overlay {
    width: 200;
    height: 200;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    text-align: center;
    font-size: 0.75rem;
}

.DonutChart__overlay__payment {
    font-size: 2rem;
    font-weight: bold;
    margin-top: 0.5rem;
}

