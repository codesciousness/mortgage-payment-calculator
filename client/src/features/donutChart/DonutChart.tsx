import React from 'react';
import { Chart } from "react-google-charts";
import './DonutChart.css';

type Expenses = 'Principal & interest' | 'Property tax' | 'Homeowner\'s insurance' | 'HOA fees' | 'Other costs';

type ColumnHeaders = [string, string];

type RowData = [Expenses, number];

const DonutChart = (): JSX.Element => {
    const columnHeaders: ColumnHeaders = ['Mortgage Related Expenses', 'Payment Amount'];
    const rowData: RowData = [
        ['Principal & interest', 11],
        ['Property tax', 2],
        ['Homeowner\'s insurance', 2],
        ['HOA fees', 2],
        ['Other costs', 7]
    ];
    const data = [columnHeaders, ...rowData];

    const total = '$2,500';

    const options = {
        pieHole: 0.65,
        is3D: false,
        pieSliceText: 'none',
        legend: 'none',
        slices: {
            0: { color: 'DeepSkyBlue' },
            1: { color: 'LightSkyBlue' },
            2: { color: 'SkyBlue' },
            3: { color: 'PowderBlue' },
            4: { color: 'PaleTurquoise' }
        },
        chartArea: {
            left: 10,
            top: 0,
            width: '95%',
            height: '100%'
        }
    };

    return (
        <table id='DonutChart' className='DonutChart'>
            <tr className='DonutChart__tableRow'>
                <td className='DonutChart__tableData'>
                    <Chart
                        chartType='PieChart'
                        width='100%'
                        height='400px'
                        data={data}
                        options={options}
                    />
                    <div className='DonutChart__overlay'>
                        <p className='DonutChart__overlay__title'>Monthly Payment</p>
                        <p className='DonutChart__overlay__payment'>{total}</p>
                    </div>
                </td>
            </tr>
        </table>
    );
};

export default DonutChart;