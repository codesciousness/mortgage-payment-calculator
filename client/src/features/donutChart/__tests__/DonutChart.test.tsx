import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { store } from '../../../app/store';
import DonutChart from '../DonutChart';

describe('<DonutChart />', () => {
    test('DonutChart component is defined', () => {
        expect(<DonutChart/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <DonutChart/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly with initial state from Redux store', () => {
        render(
            <Provider store={store}>
                <DonutChart/>
            </Provider>
        );
        expect(screen.getByText('Monthly Payment')).toBeInTheDocument();
        expect(screen.getByText('$2,328')).toBeInTheDocument();
    });
});