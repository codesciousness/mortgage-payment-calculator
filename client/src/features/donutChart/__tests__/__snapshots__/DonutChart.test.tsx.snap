// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DonutChart /> component matches the snapshot 1`] = `
<table
  className="DonutChart"
  id="DonutChart"
>
  <tbody
    className="DonutChart__tableBody"
  >
    <tr
      className="DonutChart__tableRow"
    >
      <td
        className="DonutChart__tableData"
      >
        <div
          className="DonutChart__overlay"
        >
          <p
            className="DonutChart__overlay__title"
          >
            Monthly Payment
          </p>
          <p
            className="DonutChart__overlay__payment"
          >
            $2,328
          </p>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
